pennylane/circuit_graph.py

```diff
@@ -12,305 +12,283 @@
 
 .. currentmodule:: pennylane.circuit_graph
 
 This module contains the CircuitGraph class which is used to generate a DAG (directed acyclic graph)
-based on operation and observable queues.
+representation of a quantum circuit.
 
-.. raw:: html
 
-    <h3>Summary</h3>
+Classes
+-------
 
 .. autosummary::
-    CircuitGraph
+   CircuitGraph
+   Layer
+   LayerData
 
-.. raw:: html
 
-    <h3>Code details</h3>
+CircuitGraph methods
+--------------------
+
+.. currentmodule:: pennylane.circuit_graph.CircuitGraph
+
+.. autosummary::
+   operations
+   observables
+   graph
+   wire_indices
+   ancestors
+   descendants
+   ancestors_in_order
+   descendants_in_order
+   _in_topological_order
+   layers
+   iterate_layers
+   update_node
+
+.. currentmodule:: pennylane.circuit_graph
+
+Code details
+~~~~~~~~~~~~
 """
 from collections import namedtuple
 import networkx as nx
 
 
-Command = namedtuple("Command", ["name", "op", "return_type", "idx"])
-Layer = namedtuple("Layer", ["op_idx", "param_idx"])
+Layer = namedtuple("Layer", ["ops", "param_inds"])
+"""Parametrized layer of the circuit.
+
+TODO define what a layer is
+
+Args:
+
+    ops (list[Operation]): parametrized operations in the layer
+    param_inds (tuple[int]): corresponding free parameter indices
+"""
+
+LayerData = namedtuple("LayerData", ["pre_ops", "ops", "param_inds", "post_ops"])
+"""Parametrized layer of the circuit.
 
+Args:
+    pre_ops (list[Operation]): operations that precede the layer
+    ops (list[Operation]): parametrized operations in the layer
+    param_inds (tuple[int]): corresponding free parameter indices
+    post_ops (list[Operation]): operations that succeed the layer
+"""
 
 class CircuitGraph:
-    """Represents a queue of operations and observables
-    as a directed acyclic graph.
+    """Represents a quantum circuit as a directed acyclic graph.
 
     Args:
-        queue (list[Operation]): the quantum operations to apply
-        observables (list[Observable]): the quantum observables to measure
-        parameters (dict[int->list[(int, int)]]): Specifies the free parameters
-            of the quantum circuit. The dictionary key is the parameter index.
-            The first element of the value tuple is the operation index,
+        ops (Iterable[Operation]): quantum operations constituting the circuit, in temporal order
+        variable_deps (dict[int, list[ParDep]]): Free parameters of the quantum circuit.
+            The dictionary key is the parameter index.
+            The first element of the ParDep namedtuple is the operation index,
             the second the index of the parameter within the operation.
     """
-
-    def __init__(self, queue, observables, parameters=None):
-        self._queue = queue
-        self._observables = observables
-        self.parameters = parameters or {}
+    def __init__(self, ops, variable_deps):
+        self.variable_deps = variable_deps
 
         self._grid = {}
```

Comment:
I wonder if it (1) make sense to be more explicit about `self._grid` from the get go, since the number of wires is known from the beginning and (2) to turn `self._grid` into a list instead of a dict, as we are not using any dict aspects at all here. For example,

```python
self._grid = [[]] * num_wires
```
 We could consider initializing CircuitGraph with num_wires, or determining num_wires using `ops[0]` for example (I'm in favour of the latter).

This change would simplify line `105` to be:
```python
self._grid[w].append(op)
```

And line `109` to be:
```python
for wire in self._grid
```

Reply:
There might be empty wires in the circuit, in which case the dict won't have a corresponding entry, and len(_grid) will be the number of used wires. Overall I think it doesn't matter much since _grid is essentially a temporary data structure we use to build the graph.